ldms/src/sampler/dcgm_sampler/gen-ldms-dcgm-list-fields

```diff
@@ -0,0 +1,90 @@
+#! /bin/bash
+cat << EOF
+/* generated with gen_ldms_dcgm_list_fields */
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <dcgm_agent.h>
+struct fielddef { const char *macro; int field_id; char *tag;};
+#define FIELDDEF(m) { #m, m, NULL },
+struct fielddef all_fields[] = {
+EOF
+grep DCGM_FI_ /usr/include/dcgm_fields.h |grep '^#define' |grep ' [0-9]'|grep -v + | sed -e 's/#define //' -e 's/ .*//g' -e 's/^/FIELDDEF(/' -e 's/$/)/'
+cat << EOF
+	{ NULL, 0, NULL }
+};
+
+
+#define ARRAY_SIZE(a) (sizeof(a) / sizeof(*a))
+#define _GNU_SOURCE
+
+
+static int dcgm_init()
+{
+        dcgmReturn_t rc;
+
+        rc = dcgmInit();
+        if (rc != DCGM_ST_OK) {
+                return -1;
+        }
+
+        return 0;
+}
+
+static void dcgm_fini()
+{
+        dcgmShutdown();
+}
+
+const char *typeString(int ft)
+{
+	switch (ft) {
+	case DCGM_FT_DOUBLE:
+		return "double";
+	case DCGM_FT_INT64:
+		return "int64_t";
+	case DCGM_FT_STRING:
+		return "string";
+	case DCGM_FT_TIMESTAMP:
+		return "timestamp";
+	default:
+		return "unmapped_type";
+	}
+}
+
+static void dump_dcgm_tags()
+{
+        int i;
+	printf("field_id\ttag\tmacro\ttype\tunits\n");
+        for (i = 0; i < ARRAY_SIZE(all_fields) &&
```

Comment:
Calls for values not listed in the header are specified as yielding undefined results. Calling with unsupported input is bad programming practice and, as demonstrated, unnecessary.

Reply:
Let's quote the entire header if arguing 3rd party behavior, since it appears libdcgm is not open-source:
```
DcgmFieldGetById ( unsigned short fieldId )
    Parameters

    fieldId
        IN: One of the field IDs (DCGM_FI_?) 

    Returns

    0 On Failure > 0 Pointer to field metadata structure if found.
    Description

    Get a pointer to the metadata for a field by its field ID. See DCGM_FI_? for a list of field IDs. 
```

Please note the:
```
    fieldId
        IN: One of the field IDs (DCGM_FI_?) 
```
It doesn't say 
```
    fieldId
        IN: a random number from 0 to DCGM_FI_MAX_FIELDS that may or may not be a valid field_id.
```

I check for NULL return because there are specific, valid DCGM_FI_ values that do not have an associated meta-data structure. As the code generation is trivial and the tool needs to be reliable for administrators to accept it, there's no reason to not be conservative in the coding of this tool. This is a case that is well-known to autoconf.

> I don't understand why you are making that claim. You already check for NULL returned by DcgmFieldGetById, which is in fact the documented, expected return (they call it 0) when the structure is not found (i.e. the value is not listed in the header).
> 
> Here is what the documentation says:
> 
> "Returns 0 On Failure > 0 Pointer to field metadata structure if found."

Reply:
I think that the documentation for "Results" _very_ clearly tells us what happens when there is not a "metadata structure...found" for the given number. It returns 0. If you want to get pedantic, then the "Results" documentation literally leaves no other return options other than:

    0
    field metadata structure

Literally no undefined behavior exists between those two possibilities.

The code will be _entirely reliable_ for administrators without all this unnecessary complexity.

Unless you can demonstrate an actual bug in DcgmFieldGetId(), there is no good reason thus far for all of this complexity.

Reply:
@morrone, @tom95858  I ran a test code over the loop Chris proposed( 0:DCGM_FI_MAX_FIELDS )for potential field ids using DcgmFieldGetById. This is to avoid the trivial step of extracting documented field ids from the published dcgm header.

The result runs with valgrind cleanly, but it returns meta data for undocumented fields (which may or may not be actually supported correctly by the library since they are undocumented).

This is potentially confusing to plugin users, with LDMS becoming blamed for reporting incorrect data when in fact it's just a field still in development/retirement or otherwise unheadered.

Do we want to show users the documented fieldIds only or the output from the library including unheadered values of fieldId? 
My understanding, based on the Cray network sampler cases, is that we don't collect undocumented things.

Reply:
We could always forgo the program altogether and just point users at the documentation if it really worries you.

Reply:
I want to make accessible to the configurer (before running ldms, so they can know what schema to set up at their database end) the mapping of the numbers/macros from the dcgm headers to the metric names the sampler will produce. 
The proposed utility does this, though obviously it doesn't substitute for reading more detailed documentation.
@morrone 
If you want to reduce "build complexity", I can move the dumping of this information into the usage() call. Would you prefer that?